wen-restart/src/wen_restart.rs

```diff
@@ -405,6 +415,188 @@ fn find_bankhash_of_heaviest_fork(
+        cluster_info.my_shred_version(),
+        epoch_stakes,
+        heaviest_fork_slot,
+        heaviest_fork_hash,
+        &cluster_info.id(),
+    );
+    if let Some(aggregate_record) = &progress.heaviest_fork_aggregate {
+        for (key_string, message) in &aggregate_record.received {
+            if let Err(e) = heaviest_fork_aggregate.aggregate_from_record(key_string, message) {
+                error!("Failed to aggregate from record: {:?}", e);
+            }
+        }
+    } else {
+        progress.heaviest_fork_aggregate = Some(HeaviestForkAggregateRecord {
+            received: HashMap::new(),
+            final_result: None,
+        });
+    }
+
+    let mut total_active_stake = heaviest_fork_aggregate.total_active_stake();
+    progress
+        .my_heaviest_fork
+        .as_mut()
+        .unwrap()
+        .total_active_stake = total_active_stake;
+    cluster_info.push_restart_heaviest_fork(
+        heaviest_fork_slot,
+        heaviest_fork_hash,
+        total_active_stake,
+    );
+
+    let mut progress_last_sent = Instant::now();
+    let mut cursor = solana_gossip::crds::Cursor::default();
+    let mut progress_changed = false;
+    loop {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        let start = timestamp();
+        for new_heaviest_fork in cluster_info.get_restart_heaviest_fork(&mut cursor) {
+            let from = new_heaviest_fork.from.to_string();
+            if let Some(record) = heaviest_fork_aggregate.aggregate(new_heaviest_fork) {
+                progress
+                    .heaviest_fork_aggregate
+                    .as_mut()
+                    .unwrap()
+                    .received
+                    .insert(from, record);
+                progress_changed = true;
+            }
+        }
+        let current_total_active_stake = heaviest_fork_aggregate.total_active_stake();
+        if current_total_active_stake > total_active_stake {
+            total_active_stake = current_total_active_stake;
+            progress
+                .my_heaviest_fork
+                .as_mut()
+                .unwrap()
+                .total_active_stake = current_total_active_stake;
+            progress_changed = true;
+        }
+        if progress_changed {
+            progress_changed = false;
+            let total_active_stake_seen_supermajority =
+                heaviest_fork_aggregate.total_active_stake_seen_supermajority();
+            info!(
+                "Total active stake seeing supermajority: {} Total active stake: {} Total stake {}",
+                total_active_stake_seen_supermajority,
+                heaviest_fork_aggregate.total_active_stake(),
+                total_stake
+            );
+            let can_exit = total_active_stake_seen_supermajority as f64 * 100.0
+                / total_stake as f64
+                > wait_for_supermajority_threshold_percent as f64;
```

Comment:
Does this also need to accommodate the 5% malicious

Reply:
Yes, the 5% can choose to exit in the middle.